pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -531,7 +539,17 @@ public boolean commitSegment(String rawTableName, final String committingSegment
 
     // TODO Introduce a controller failure here for integration testing
 
-    updateHelixIdealState(realtimeTableName, newInstances, committingSegmentNameStr, newSegmentNameStr);
+    // When multiple segments of the same table complete around the same time it is possible that
+    // the idealstate udpate fails due to contention. We serialize the updates to the idealstate
+    // to reduce this contention. We may still contend with RetentionManager, or other updates
```

Comment:
Speaking aloud, we should have internal queue and let single thread manage updates to ideal state. We can batch updates, if helix supports batching, reducing failure window.